jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java

```diff
@@ -376,6 +376,13 @@ private boolean parseRequestBuffer()
     @Override
     public void onCompleted()
     {
+        // If we are fill interested, then a read is pending and we must abort
+        if (isFillInterested())
+        {
+            LOG.warn("Pending read in onCompleted {} {}", this, getEndPoint());
+            abort(new IllegalStateException());
+        }
```

Comment:
Do we want to continue after this abort? Should we not return?

Reply:
Probably, but need to check there are no other required actions afterwards

Reply:
I've just skipped the 101 upgrade, but will keep the rest of the method as the resets and recycles look good for cleanup.